hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestFinalizeBlock.java

```diff
@@ -0,0 +1,355 @@
+
+    builder.setFinalizeBlock(ContainerProtos.FinalizeBlockRequestProto
+        .newBuilder().setBlockID(blockId).build());
+
+    XceiverClientManager xceiverClientManager = new XceiverClientManager(conf);
+    XceiverClientSpi xceiverClient =
+        xceiverClientManager.acquireClient(pipeline);
+
+    ContainerProtos.ContainerCommandRequestProto request = builder.build();
+    ContainerProtos.ContainerCommandResponseProto response =
+        xceiverClient.sendCommand(request);
+
+    Assert.assertTrue(response.getFinalizeBlock()
+        .getBlockData().getBlockID().getLocalID()
+        == omKeyLocationInfoGroupList.get(0)
+        .getLocationList().get(0).getLocalID());
+    Assert.assertTrue(((KeyValueContainerData)
+        getContainerfromDN(cluster.getHddsDatanodes().get(0),
+        containerId.getId()).getContainerData())
+        .getFinalizedBlockSet().size() == 1);
+
+    try {
+      cluster.restartHddsDatanode(0, false);
+    } catch (Exception e) {
+    }
+
+    // After restart DN, finalizeBlock should be loaded into memory
+    Assert.assertTrue(((KeyValueContainerData)
+        getContainerfromDN(cluster.getHddsDatanodes().get(0),
+        containerId.getId()).getContainerData())
+        .getFinalizedBlockSet().size() == 1);
+  }
+
+  @Test
+  public void testFinalizeBlockClearAfterCloseContainer()
+      throws IOException {
+    String keyName = UUID.randomUUID().toString();
+    // create key
+    createKey(keyName);
+
+    ContainerID containerId = cluster.getStorageContainerManager()
+        .getContainerManager().getContainers().get(0).containerID();
+
+    OmKeyArgs keyArgs = new OmKeyArgs.Builder().setVolumeName(volumeName)
+        .setBucketName(bucketName).setKeyName(keyName).setDataSize(0)
+        .build();
+    List<OmKeyLocationInfoGroup> omKeyLocationInfoGroupList =
+        cluster.getOzoneManager().lookupKey(keyArgs).getKeyLocationVersions();
+
+
+    ContainerInfo container = cluster.getStorageContainerManager()
+        .getContainerManager().getContainer(containerId);
+    Pipeline pipeline = cluster.getStorageContainerManager()
+        .getPipelineManager().getPipeline(container.getPipelineID());
+
+
+    final ContainerProtos.ContainerCommandRequestProto.Builder builder =
+        ContainerProtos.ContainerCommandRequestProto.newBuilder()
+            .setCmdType(ContainerProtos.Type.FinalizeBlock)
+            .setContainerID(container.getContainerID())
+            .setDatanodeUuid(cluster.getHddsDatanodes()
+            .get(0).getDatanodeDetails().getUuidString());
+
+    final ContainerProtos.DatanodeBlockID blockId =
+        ContainerProtos.DatanodeBlockID.newBuilder()
+            .setContainerID(container.getContainerID()).setLocalID(
+                omKeyLocationInfoGroupList.get(0)
+            .getLocationList().get(0).getLocalID())
+            .setBlockCommitSequenceId(0).build();
+
+    builder.setFinalizeBlock(ContainerProtos.FinalizeBlockRequestProto
```

Comment:
This finalize block request build process can be extracted into a individual function. The process is the same for all tests in this class.